### PR TITLE
Rewrited version of plugin

### PR DESCRIPTION
**Pull request by lanwen**
contains #3 

simplify UI/UX to single page. Refactor some internal logic to just always rewrite actual list of sites

2016-01-08 02-12-50 manage updatesites jenkins

[Review on Reviewable](https://reviewable.io/reviews/jenkinsci/update-sites-manager-plugin/4)

### CODE REVIEW

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/internal/ExtendedCertJsonSignValidatorTest.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    public void shouldAddCustomCertToTrustAnchors() throws Exception {
+        String cert = IOUtils.toString(getClass().getClassLoader()
+                .getResourceAsStream("jp/ikedam/jenkins/plugins/updatesitesmanager/" +
+                        "ManagedUpdateSiteJenkinsTest/caCertificate.crt"), Charsets.UTF_8);
+        CertificateFactory cf = CertificateFactory.getInstance("X509");
+
+        Set<TrustAnchor> test = new ExtendedCertJsonSignValidator("test", cert).loadTrustAnchors(cf);
+        assertThat(test, hasSize(1));
+    }
+}
```
Comment (KostyaSha) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/internal/ExtendedCertJsonSignValidatorTest.java]: NL

File: pom.xml [hunk 2]
```diff
@@ -55,16 +55,17 @@
...
-    </dependency>
-    <dependency>
-      <groupId>commons-lang</groupId>
-      <artifactId>commons-lang</artifactId>
-      <version>[2.3,]</version>
-    </dependency>
-  </dependencies>
+    
+    <build>
+        <plugins>
```
Comment (oleg-nenashev) [inline pom.xml]: NIT: it does not follow the indenting of other sections

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSite.java [hunk 3]
```diff
@@ -216,102 +153,13 @@ public boolean hasUpdates()
...
-    /**
      * Returns all DescribedUpdateSite classes registered to Jenkins.
      * 
      * @return the list of Descriptor of DescribedUpdateSite subclasses.
      */
-    static public DescriptorExtensionList<DescribedUpdateSite,DescribedUpdateSite.Descriptor> all()
+    static public DescriptorExtensionList<DescribedUpdateSite, DescribedUpdateSite.Descriptor> all()
     {
-        return Jenkins.getInstance().<DescribedUpdateSite,DescribedUpdateSite.Descriptor>getDescriptorList(DescribedUpdateSite.class);
+        return Jenkins.getInstance().getDescriptorList(DescribedUpdateSite.class);
```
Comment (oleg-nenashev) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSite.java]: getActiveInstance()

File: src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/sidepanel.jelly [hunk 4]
```diff
@@ -26,8 +26,7 @@ THE SOFTWARE.
   <l:header />
   <l:side-panel>
     <l:tasks>
-      <l:task icon="images/24x24/up.png" href=".." title="${%Back to Manage Jenkins}" contextMenu="false" />
-      <l:task icon="images/24x24/new-package.png" href="_add" title="${%Add New Site}" />
+      <l:task icon="images/24x24/up.png" href="../manage" title="${%Back to Manage Jenkins}" contextMenu="false" />
```
Comment (oleg-nenashev) [inline src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/sidepanel.jelly]: Would be better to use an absolute link. Just to avoid issue s with UI restructuring

Comment (lanwen) [inline src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/sidepanel.jelly]: how to construct that in jelly?

**Comment by ikedam** [comment]
Really big change.
Would you tell me followings?
* The motivation of this change.
* The outline if the change.

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/ManagedUpdateSite.java [hunk 5]
```diff
@@ -54,10 +51,7 @@
  *   <li>can set a CA certificate for the signature of the site.</li>
  * </ul>
  * 
- * The CA certificate is written to a temporary file under ${JENKINS_HOME}/update-center-rootCAs/ .
- * I think it is a better implementation to override UpdateSite#verifySignature,
- * but it is private method.
- * Re-implementing doPostBack results to re-implement whole the UpdateSite.
+ * The CA certificate is written as additional trust anchor dynamically
```
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/ManagedUpdateSite.java]: This change should be in #3.

**Comment by lanwen** [comment]
Motivation:
1. I cant see what exactly under the hood for any ID - i should click.
2. I cant change anything without click to id
3. I can't change the order if UCs
4. Code contains lot of endpoint with complex logic - lot of troubles to maintain
5. There is a ready-to-use component which already used for clouds, install tools etc.

Outline:
1. Simply use hetero-list to show describables
2. Only one endpoint which get all custom UCs and persists them (combined with not managed)
3. Its not need custom describable to show not managed sites as we don't want to edit them - so simply show them inline
4. Use some guava to simplify filtering and transforming
5. Remove editable property as of if anybody wants to setup custom describable, but don't allow this plugin to manage it - they can simply write own class

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/ManagedUpdateSite.java (hunk 5 above)
Comment (lanwen) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/ManagedUpdateSite.java]: yep, forgot about it in #3

**Comment by ikedam** [comment]
Thanks for details.
I basically agree this change. Please let me have some more time to review details.

I also plan to additional changes after merging this change to simplify this plugin (e.g. removing `Serializable`, removing `canCreateNewSite`) .

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java [hunk 6]
```diff
@@ -23,274 +23,205 @@
...
      * @see hudson.model.Action#getDisplayName()
      */
     @Override
-    public String getDisplayName()
-    {
+    public String getDisplayName() {
         return Messages.UpdateSitesManager_DisplayName();
     }
-    
+
```
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java]: I highly recommend you to disable auto formatter of your IDE (especially when contributing OSS).
Unrelated changes makes it hard for reviewers to see the actual change you want and to track changes with `git blame`.
(I don't think you need to fix them and remake this request)

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java [hunk 7]
```diff
@@ -23,274 +23,205 @@
...
+    public HttpResponse doUpdate(@Sites List<UpdateSite> managed) throws ServletException, IOException, FormException {
+        List<UpdateSite> newSitesList = newArrayList(Iterables.concat(getNotManagedUpdateSiteList(), managed));
+
+        shouldNotContainDuplicatedIds(newSitesList);
+        shouldNotContainBlankIds(newSitesList);
+
+        Jenkins.getInstance().getUpdateCenter().getSites().replaceBy(newSitesList);
+        Jenkins.getInstance().getUpdateCenter().save();
+
+        return HttpResponses.redirectToDot();
```
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java]: Better to redirect to "Manage Jenkins".

File: src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly [hunk 8]
```diff
@@ -22,39 +22,45 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
+        <l:header/>
+        <l:main-panel>
+            <h1>
+                <img src="${imagesURL}/48x48/${it.iconFileName}" alt=""/>${it.displayName}
+            </h1>
+
+            <h2>
+                Not managed update sites
+            </h2>
+            <div class="warning">${%This sites cannot be configured with UpdateSitesManager. Please visit the configuration page dedicated for this site.}</div>
```
Comment (ikedam) [inline src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly]: This -> These

File: src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly [hunk 9]
```diff
@@ -22,39 +22,45 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
+
+            <ul>
+                <f:repeatable var="site" noAddButton="true" items="${it.getNotManagedUpdateSiteList()}">
```
Comment (ikedam) [inline src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly]: JFYI: Groovy allows you to use `notManagedUpdateSiteList` instead of `getNotManagedUpdateSiteList()`

File: src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly [hunk 10]
```diff
@@ -22,39 +22,45 @@ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
...
-    <a id="add-new-site-link" href="_add" title="${%Add New Site}">${%Add New Site}</a>
-  </l:main-panel>
-</l:layout>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:l="/lib/layout" xmlns:f="/lib/form">
+    <!--
+      shown when accessed /updatesites/
+    -->
+    <l:layout title="${it.displayName}" permission="${app.ADMINISTER}" norefresh="true">
+        <st:include page="sidepanel.jelly"/>
+        <l:header/>
```
Comment (ikedam) [inline src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly]: You no longer need l:header.

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java [hunk 11]
```diff
@@ -24,388 +24,122 @@
...
  * Tests for DescribedUpdateSite, concerned with Jenkins.
  */
-public class DescribedUpdateSiteJenkinsTest extends HudsonTestCase
-{
-    public static class DescribedUpdateSiteForConfigureTest extends DescribedUpdateSite
-    {
+public class DescribedUpdateSiteJenkinsTest {
+
+    @Rule
+    public JenkinsRule jRule = new JenkinsRule();
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: JFYI: `JenkinsRule` with `@Rule` launches Jenkins for each tests.
Using `@ClassRule` saves test execution time.

Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: (already applied in `ManagedUpdateSiteJenkinsTest.java`)

**Comment by ikedam** [comment]
Motivation:
6. Update sites are saved into one configuration file, and it's easy to understand that there's one configuration page for one configuration file.

> 5.There is a ready-to-use component which already used for clouds, install tools etc.

I could not get what this means.
Would you let me know more details?

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java (hunk 11 above)
Comment (KostyaSha) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: Yes, that what we said to you about rules few month ago ;)

Comment (lanwen) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: Yep, i know about it. But newer versions of jenkins tuned up this rule and restart is very quick. So if you have less than 5 tests overhead of Rule is a few seconds

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java (hunk 6 above)
Comment (lanwen) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java]: yep, sorry for that. This plugin uses another style than most of plugins, and i forgot about turning that off

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java (hunk 7 above)
Comment (lanwen) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java]: I think that it should be some kind of verification for user that all saved without any problem. So i don't know on how to make *apply* btn and make combined behaviour in the *save* btn

File: src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly (hunk 9 above)
Comment (lanwen) [inline src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly]: so can i convert the view to groovy?

**Comment by lanwen** [comment]
>Would you let me know more details?
Its all about `hetero-list` tag (or `repeatable` as more common)

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java (hunk 7 above)
Comment (KostyaSha) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java]: throw `FormException` in case of error?

Comment (lanwen) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java]: thats already so. But i prefer the some kind of warning that all ok (as done in global config on apply) or the refresh (as done here)

File: src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly (hunk 9 above)
Comment (ikedam) [inline src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly]: Sorry I wrote a wrong comment.
Jelly allows groovy-like style `notManagedUpdateSiteList` instead of `getNotManagedUpdateSiteList()`.

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java (hunk 7 above)
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java]: You can put apply button with `f:apply`.
No additional implementation is required to Java code.

File: src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly (hunk 9 above)
Comment (lanwen) [inline src/main/resources/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager/index.jelly]: ah, ok, will change

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/internal/IgnoreNotPOST.java [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+/**
+ * Just redirects to url of sites manager if not POST
+ *
+ * @author lanwen (Merkushev Kirill)
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+@Documented
+@InterceptorAnnotation(IgnoreNotPOST.Processor.class)
+public @interface IgnoreNotPOST {
```
Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/internal/IgnoreNotPOST.java]: How about existing `@RequirePOST`?

Comment (ikedam) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/internal/IgnoreNotPOST.java]: (replying to myself)
`@RequirePOST` displays an error page with a button to have users POST without any parameters, which should result users lose all managed update sites.
It is completely useless and we need another implementation (like this `@IgnoreNotPost`).

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java (hunk 7 above)
Comment (lanwen) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManager.java]: `f:apply` shows error popup after click

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java (hunk 11 above)
Comment (lanwen) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: also as we manage static state of whole jenkins, it much more safer to use rule instead of class rule

File: src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/internal/IgnoreNotPOST.java (hunk 12 above)
Comment (lanwen) [inline src/main/java/jp/ikedam/jenkins/plugins/updatesitesmanager/internal/IgnoreNotPOST.java]: don't think users lose managed update sites with that annotation (it throws same IvocationException), but error page with `@RequirePOST` is ugly and can scare end user if something goes wrong

**Comment by ikedam** [comment]
Note for me: Mapping of tests:

* DescribedUpdateSiteJenkinsTest

    | Old test        | New test      |
    |:----------------|:--------------|
    | testDoConfigure | (removed, replaced with UpdateSitesManagerJenkinsTest#shouldSubmitSites)     |
    | testDoDelete    | (removed)     |
    | testPrivilege   | testPrivilege |
    |                 | shouldShowAllDescribedSitesAsManaged |
    |                 | shouldRedirectOnGetReqOfUpdate |

* DescribedUpdateSiteSimpleTest

    | Old test                             | New test                          |
    |:-------------------------------------|:----------------------------------|
    | testDescribedUpdateSite / trimmed    | shouldTrimUrlAndId                |
    | testDescribedUpdateSite / null       | shoulHandleNull                   |
    | testGetUpdateSite                    | (removed)                         |
    | testDescriptorDoCheckIdOk            | testDescriptorDoCheckIdOk         |
    | testDescriptorDoCheckIdError / null  | testDescriptorDoCheckIdErrorNull  |
    | testDescriptorDoCheckIdError / empty | testDescriptorDoCheckIdErrorEmpty |
    | testDescriptorDoCheckIdError / blank | testDescriptorDoCheckIdErrorBlank |
    | testDescriptorDoCheckUrlOk           | testDescriptorDoCheckUrlOk        |
    | testDescriptorDoCheckUrlError        | testDescriptorDoCheckUrlError     |

* DescribedUpdateSiteWrapperJenkinsTest
    * removed
* DescribedUpdateSiteWrapperSimpleTest
    * removed
* ManagedUpdateSiteJenkinsTest
    * (Skipped as already reviewed in #3, #5).
* UpdateSitesManagerJenkinsTest

    | Old test                                         | New test                         |
    |:-------------------------------------------------|:---------------------------------|
    | testManagementLink                               | shouldExistsLinkToManager        |
    | testNoUpdateSitesInList                          | (removed)                        |
    | testManyUpdateSitesInList                        | (removed)                        |
    | testGetUpdateSiteList / no update sites          | shouldNotReturnAnyUCIfEmptyList  |
    | testGetUpdateSiteList / many update sites        | shouldReturnBothManagedUnmanaged |
    | testGetUpdateSiteList_DescribedUpdateSiteWrapper | (removed)                        |
    | testGetUpdateSiteDescriptorList                  | testGetUpdateSiteDescriptorList  |
    |                                                  | shouldSubmitSites                |
    | testGetDynamic                                   | (removed)                        |
    | testDo_add                                       | (removed)                        |
    | testDo_addWithMultiUpdateSite                    | (removed)                        |
    | testDo_addError / Post without id                | shouldReturn400OnBlankId         |
    | testDo_addError / Post with an duplicated id     | shouldReturn400OnDuplicatedId    |
    | testPrivilege                                    | testPrivilege                    |

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java [hunk 13]
```diff
@@ -24,388 +24,122 @@
...
+            public String getDisplayName() {
                 return "DescribedUpdateSiteForConfigureTest";
             }
         }
     }
-    public void testDoConfigure() throws Exception
-    {
+
+    @Test
+    public void shouldShowAllDescribedSitesAsManaged() throws Exception {
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: Looks duplicating `UpdateSitesManagerJenkinsTest#shouldReturnBothManagedUnmanaged`

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java [hunk 14]
```diff
@@ -24,388 +24,122 @@
...
-            {
-                assertEquals("This request must be rejected with 400 Bad Request", 400, e.getStatusCode());
-            }
-            assertEquals("UpdateSite must not be deleted.", initialSize, Jenkins.getInstance().getUpdateCenter().getSites().size());
-        }
+        UpdateSitesManager manager = new UpdateSitesManager();
+
+        assertThat("should register described UC only", manager.getManagedUpdateSiteList(), hasSize(1));
+        assertThat("should save class of UC",
+                manager.getManagedUpdateSiteList().get(0), (Matcher) isA(DescribedUpdateSiteForConfigureTest.class));
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: You can test the instance (not its class) using `JenkinsRule#assertEqualDataBoundBeans`.

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java [hunk 15]
```diff
@@ -24,204 +24,90 @@
...
-            String url = "hogehoge";
-            assertEquals(
-                "non url",
-                FormValidation.Kind.ERROR,
-                descriptor.doCheckUrl(url).kind
-            );
-        }
+
+    @Test
+    public void testDescriptorDoCheckUrlError() {
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java]: Inconsistent policy with `testDescriptorDoCheckIdErrorXxxx` (one test method for one test condition).

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManagerJenkinsTest.java [hunk 16]
```diff
@@ -23,624 +23,213 @@
...
+        // configure
+        HtmlForm form = wc.goTo(UpdateSitesManager.URL).getFormByName("sitesForm");
+        HtmlPage submit = jRule.submit(form);
+
+        assertThat("should be on same url after submit",
+                submit.getWebResponse().getUrl().toString(), endsWith(UpdateSitesManager.URL + "/"));
+
+        assertThat("should see all sites", jRule.getInstance().getUpdateCenter().getSites(), hasSize(2));
+        assertThat("should see custom site",
+                jRule.getInstance().getUpdateCenter().getSites(), hasItem(isA(ManagedUpdateSite.class)));
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManagerJenkinsTest.java]: Compare the attributes of each sites. (`JenkinsRule#assertEqualDataBoundBeans` will help you).
It allows you to test your jelly files are correct.

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManagerJenkinsTest.java [hunk 17]
```diff
@@ -23,624 +23,213 @@
...
+    public void testPrivilege() throws Exception {
+        JenkinsRule.WebClient wcUser = jRule.createWebClient();
+        wcUser.login("user", "user");
+
+        ex.expect(FailingHttpStatusCodeException.class);
+        ex.expectMessage(containsString("403"));
+        wcUser.goTo(String.format("%s/update", UpdateSitesManager.URL));
+    }
+
+    static public class TestUpdateSite extends DescribedUpdateSite {
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManagerJenkinsTest.java]: No longer used.

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManagerJenkinsTest.java [hunk 18]
```diff
@@ -23,624 +23,213 @@
...
-            
-            @Override
-            public String getDisplayName()
-            {
-                return "Provided for testing purpose";
-            }
-        }
-    }
-    
-    static public class SingletonUpdateSite extends DescribedUpdateSite
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/UpdateSitesManagerJenkinsTest.java]: Remove also resource files (placed in test/resources).

**Comment by ikedam** [comment]
I finished reviewing the whole change.
Sorry for having you wait for so long time.

I will merge #5 and you might need to rebase commits.

Required fixes for this request:
* Fix the page transition when configured update sites.
    * Though I can get what you want with the current implementation, I don't like it as following reasons:
        * We should do as UI / UX of Jenkins core not to confuse users.
        * Jenkins core provides <f:apply> for that purpose.
            * I expect you to see why <f:apply> doesn't work as expected.
            * I can take over it if you don't want to do.
* Fix some tests as commented before.

I also plan following changes after merging this request.
These are issues in existing codes and I don't think issues of this request.
Of course, you can fix them in this request you like.

* Change links to absolute
* Remove `Serializable`
    * I no longer plan to preserve binary compatibilities in the next release.
* Remove `DescribedUpdateSite.Descriptor#canCreateNewSite`
* More internationalization for `UpdateSitesManager/index.jelly`

**Comment by ikedam** [comment]
Possible regression:
* Sites with `canCreateNewSite=false` and `isEditable=false` will be no longer supported.

Planned action:
* Don't care.
* I don't think that regression actually affects anyone. There's no official plugins depending on update-sites-manager-plugin.
* I also doubt `canCreateNewSite` and `isEditable` is useful for developers who want to derive `DescribedUpdateSite`
* I will provide alternate feature in future release if someone reports regressions.

**Comment by lanwen** [comment]
Thanks for review! Will implement proposed notes

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java (hunk 13 above)
Comment (lanwen) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: no, the difference is the type of sites - here both has the same type

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java (hunk 15 above)
Comment (lanwen) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java]: i was too lazy to rewrite this part :)

**Comment by lanwen** [comment]
@ikedam rebased and seems fixed all notes (but i can miss something while rebasing...)

File: pom.xml [hunk 19]
```diff
@@ -55,4 +55,23 @@
       <url>http://repo.jenkins-ci.org/public/</url>
     </pluginRepository>
   </pluginRepositories>
+    
+    <dependencies>
```
Comment (ikedam) [inline pom.xml]: As pointed before, please keep consistent indentations.

File: pom.xml [hunk 20]
```diff
@@ -55,4 +55,23 @@
...
+            <version>4.12</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-all</artifactId>
+            <version>1.3</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (ikedam) [inline pom.xml]: Why do you need those dependencies?

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java (hunk 13 above)
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteJenkinsTest.java]: Hmm... I actually can't get the difference. I can't imagine a case only one of them fails. (both tests will fail if something wrong).
But I think it's OK if you like it.

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java (hunk 15 above)
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java]: Don't be lazy here, please.
You refactored tests as you thought the existing ones were bad, did you?
(And I agree with that. I didn't know much about tests and I now believe your new ones are desirable)
Someone should refactor it again in future.

**Comment by ikedam** [comment]
My new planned task after merging this request:
* To see why f:apply doesn't work as expected.

**Comment by ikedam** [comment]
I want you not have squashed the rebased commit and the commit for the additional change.
It would make me easy to review your additional change.

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java (hunk 15 above)
Comment (lanwen) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java]: okay, will do it in a few days

File: pom.xml (hunk 20 above)
Comment (lanwen) [inline pom.xml]: It may be temporary bug with my IDE, but it can't find `Matchers.hasSize()` matcher. Firstly i've think that it not included to non-lts release, but now i don't know. Will check it once more with clean pull.

**Comment by lanwen** [comment]
okay, will also rewrite test with jetty-server.

**Comment by ikedam** [comment]
> You can put apply button with f:apply.
> No additional implementation is required to Java code.

This was wrong.
I'm really sorry.
I found the page after transition was shown in the popup when only introducing f:apply to jelly.

I have to use `FormApply` in the server side.
* https://github.com/jenkinsci/jenkins/blob/jenkins-1.609/core/src/main/java/hudson/util/FormApply.java
* https://github.com/jenkinsci/jenkins/blob/jenkins-1.609/core/src/main/java/jenkins/model/Jenkins.java#L2863

Let me introduce it in following pull requests.
That allows us update the configuration without transition.

**Comment by lanwen** [comment]
@ikedam updated

File: pom.xml [hunk 21]
```diff
@@ -55,4 +55,24 @@
...
   </pluginRepositories>
+
+  <dependencies>
+    
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.12</version>
+      <scope>test</scope>
+    </dependency>
```
Comment (ikedam) [inline pom.xml]: (Note for me)
junit-4.11 will be used without this.
(I'm not sure who defines that version)

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java [hunk 22]
```diff
@@ -23,205 +23,93 @@
...
-                FormValidation.Kind.ERROR,
-                descriptor.doCheckId(id).kind
-            );
-        }
+
+    @Test
+    @DataProvider(value = {
+            "empty,",
+            "blank, ",
+            "null value,null"
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/DescribedUpdateSiteSimpleTest.java]: (Note for me)
`DataProvider` converts "null" to `null`.
https://github.com/TNG/junit-dataprovider/wiki/Features

File: src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/testext/WebServerRecipe.java [hunk 23]
```diff
@@ -0,0 +1,127 @@
...
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.lang.StringUtils;
+import org.jvnet.hudson.test.JenkinsRecipe;
+import org.jvnet.hudson.test.JenkinsRule;
+import org.mortbay.jetty.HttpConnection;
+import org.mortbay.jetty.Server;
+import org.mortbay.jetty.bio.SocketConnector;
+import org.mortbay.jetty.handler.AbstractHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (ikedam) [inline src/test/java/jp/ikedam/jenkins/plugins/updatesitesmanager/testext/WebServerRecipe.java]: (Note for me)
slf4j is included via jenkins-war.
jenkins-core starts to use slf4j since 1.628.

**Comment by ikedam** [comment]
Thanks for the update.

Really appreciate for the refactoring!
Looks perfect.
Let me have more time to make some additional changes before making a new release.

File: pom.xml (hunk 21 above)
Comment (lanwen) [inline pom.xml]: junit 4.11 defined in parent of parent pom

**Comment by ikedam** [comment]
Released update-sites-manager-2.0.0.
It will be available in the update center in a day.

Thanks for your fixes and refactoring.

**Comment by lanwen** [comment]
thanks for review and release!
